runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/ControlClientPool.java

```diff
@@ -17,16 +17,12 @@
  */
 package org.apache.beam.runners.fnexecution.control;
 
-import org.apache.beam.sdk.fn.function.ThrowingConsumer;
-import org.apache.beam.sdk.util.ThrowingSupplier;
-
 /** Control client pool that exposes a source and sink of control clients. */
-public interface ControlClientPool<T extends InstructionRequestHandler> {
+public interface ControlClientPool {
 
   /** Source of control clients. */
```

Comment:
<!--thread_id:cc_182896324_t; commit:1fd121da1417624b3b84f0300648251da64b9cb5; resolved:1-->
<!--section:context-quote-->
> **jkff** wrote:
> Generally, why the higher-order functions here: why not have ControlClientPool have a function for getting a client, and a function for adding a client to the pool? That would match how other pool-like abstractions typically work.

<!--section:body-->
I want to minimize the surface area that consumers of these endpoints can interact with. We previously bundled everything into a single pool that could be read from or written to, but it made separation of responsibilities very unclear and the APIs hard to work with.

In general, the owner of the pool will be separate from the client producers and consumers. With this interface, the owner is the _only_ one with access to the client pool itself. The producer of incoming control connections (in practice, the gRPC service) should only have access to an object sink into which it feeds clients. The consumer (for example, the docker factory) only has access to the source of incoming clients.